### ISSUE TITLE
release coordination between k2-tools, k2 and k2cli

### ISSUE THREAD

**Issue by coffeepac** [samsung-cnct/kraken-lib#376]
the k2-tools, k2 and k2cli projects should have clear versioning strategies.  this strategy should have the following characteristics:
- versions should be in semver
- versions should be pushed by CNCT operated CI
- nothing should use a :latest tag by default

**Comment by coffeepac** [comment]
There are two flows of releases here:
1. CI flow, ie a new k2-tools build should trigger a k2 build which, if successful, should trigger a k2cli build
2. named release build

The first flow is fairly easy, for every successful PR merged, kick off a downstream build.  This will quickly smoke out issues without waits.  This should be added to the individual job issues.  \<add issues for all three here>  These can all use the :latest tag.

For the second flow I am less sure of the process.  A quick search of "jenkins build tag" shows that building tags is slightly annoying.  It might be easier to push tags from jenkins after its been built (GitPublisher seems to be the key here).  That would suggest the following general workflow for each project:
- a human decides to cut a new release
- click <build now> on the specified build
- this opens a parameterized page that allows for selection of source branch (default to master) and revision to build (default to x.x.x+1, ie increment patch level by one).
- click 'submit'

this will build and push the project to wherever it goes (quay or GitHub).

For k2 and k2cli, this implies there is an additional field on their build pages which is the tag of the upstream source (k2-tools for k2 and k2 for k2cli) to use.  This will default to the latest tag available.  

For k2cli this will mean changing the code to have the k2 default image version be a build-time variable (shouldn't be hard).  
For k2 this will mean changing the Dockerfile to have the FROM value be a parameter (might be tricky, but easy enough to push the Dockerfile through a filter that swaps the image tag from latest to \<whatever>).

This will allow us to rev the version of each component independently and then easily compose them together for releases.  

ToDo:
- discover if this tag default work possible in Jenkins right now
- consider using JJB (or other jenkins job templating engine) to supply the parameterization
- fill in missing issue list above

**Comment by coffeepac** [comment]
Discussion point:  re-write default config file to pin coreOS version as part of release job.  Would this need to be pushed back into the github repo or would it be sufficient to make a note in some kind of auto-generated CHANGELOG?

**Comment by davidewatson** [comment]
Why limit pinned CoreOS versions to released versions? If there is a CI job which tests each release, each passing CoreOS version bump can/should push the latest tested version into the default config.

**Comment by coffeepac** [comment]
a. would need to create a job that watches for a new AMI across all plaltforms we use CoreOS (overhead, easily soluable, maybe a good idea in general)
b. would require developers fetch an updated version of k2 for every CoreOS release instead of lazily keeping behavior as it is now

I'm not sure if those are compelling but its what I'm seeing for now.

**Comment by davidewatson** [comment]
a) Maybe each PR should grab the latest CoreOS release and use it for testing. Then we can rely on the fact that k2 is under active developement to ensure that we continue to CI new CoreOS releases lazily.

b) Why would developers need to pull the latest CoreOS release? In most cases, CI can do that for them.

**Comment by coffeepac** [comment]
a. fair point!  was planning on having the PR/latest stream pull latest anyway.  we're doing enough dev for now to cover it

b. local testing before it gets to CI.  CI is more reliable but smoking things out before it gets there is usually a good idea, especially for longer feature work

**Comment by coffeepac** [comment]
tagging default topic:  there is a plugin to do exactly this:  https://wiki.jenkins.io/display/JENKINS/Git+Parameter+Plugin

**Comment by coffeepac** [comment]
created a test jenkins instance:  https://common-jenkins-test.kubeme.io/

installed a new plugin:  https://wiki.jenkins.io/display/JENKINS/Git+Parameter+Plugin

going to build some test jobs for this.  if it works well, I'll update the base image and then install (Manully) to production.  Manual install is a problem but we'll be handling it later.

**Comment by coffeepac** [comment]
got a sample job built but I've done something wrong with the refspec.  I've also discovered I will want to probably get this done for one of the repos (k2-tools?) to work out the exact operations for interacting with github releases.  going through github releases will generate the tag and do some other stuff.  I think we will still end up doing a manual build (maybe?) because jenkins will only build each tag once.  maybe this is okay if we don't run validation against the build but right now validation sometimes fails because ¯\_(ツ)_/¯

**Comment by coffeepac** [comment]
sample job is running and only doing the release stuff.  that is a temporary thing as I would want to run validation steps against a release before we push it.

however:  this is working.  going to do some more testing/working tomorrow to make sure all of the above can be done/set a template for them.

**Comment by coffeepac** [comment]
the job type I chose relies on a plugin that has security problems and has been pulled.  the multibranch pipeline doesn't allow for easily setting build parameters from the jenkins console.  this means I need to change how I was going to do this.

options:
- create release branches (do not want)
- build when tags are created

I'm going to do a bit more poking to see if I can sort this out in some other fashion before I start heading down the 'build when a tag is created' approach.

**Comment by leahnp** [comment]
@coffeepac a few questions, sorry if you already addressed these concerns - I tried my best to follow the ideas in the conversation so far.

- What warrants a new release?
- Are we creating a bot-github-account to cut releases to github? How do passwords work with that situation?
- Is 'rolling back' a release a thing? Should there be a semi-automated or automated way to do that?
- Can you link to the 'sample job' you refer to in the above comments?
- Do we need CI to test that the versions we say work together (between k2, k2-tools & k2cli) really do work together? Also if we create the tests as we kick off this venture of versioning everything, the tests might inform our decisions.

**Comment by coffeepac** [comment]
- What warrants a new release:  originally this was going to be whenever we completed a milestone, but we will probably move to monthly or bi-weekly releases until we have a better understanding of our velocity

- Are we creating a bot-...:  yes.  there will be several service accounts to support creating releases in github and in quay.  we will create kubernetes `secrets` manually and load them into the job at runtime

- Is 'rolling back' a release...: not for now.  I would strongly advocate we fix things forward like Kubernetes does.  ie:  1.6.5 had some critical flaw so 1.6.6 was released shortly after but 1.6.5 remains available

- Sample Job:  I deleted the job as it has security issues that won't be resolved upstream and I lack the time to resolve them myself.  Sorry!  Will post the job when I've recreated it without relying on security issued plugins

- CI for checking versions:  yes.  the publishing job will do a full CI run before it publishes the new version.  my current idea is to use the CI jobs from the existing repos, just run through those with the pinned versions.

**Comment by coffeepac** [comment]
so it been long enough that I don't remember which job type that was causing problems.  I've started trying the basic freestyle job and altering the workflow to be:
- a user at the command line puts a tag on the master branch at a specific commit
- that user then manually kicks off this job to push the tagged version wherever its going

This could probably be automated via dependent jobs so this seems like a fair thing to do.

**Comment by joejulian** [comment]
Regardless of the release pacing, it would be good to tag `$new_k2_name` when `$new_k2cli_name` is released, just for tracking purposes.

**Comment by coffeepac** [comment]
Gah, thought I had it but the var env.BRANCH_NAME is only being set in the multibranch pipeline jobs, not the regular pipeline job I'm using for this.  

will pick it up tomorrowl

**Comment by coffeepac** [comment]
if this PR works:  https://github.com/samsung-cnct/k2-tools/pull/117 (it should) I will be able to knock out the next step, k2, quickly.  

k2cli will take a bit more effort as it'll be a build-time switch to use latest or a tagged version of k2

**Comment by coffeepac** [comment]
the Dockerfile syntax I want to use isn't supported by the version of docker running in our production cluster.  I will need to do some `sed` work here instead.

**Comment by coffeepac** [comment]
K2 release PR:  https://github.com/samsung-cnct/k2/pull/763

**Comment by coffeepac** [comment]
k2cli release PR: https://github.com/samsung-cnct/k2cli/pull/144

**Comment by coffeepac** [comment]
what's left:  
 - building the k2cli release path jenkins job
 - making sure everything works

**Comment by coffeepac** [comment]
k2 release path works, k2-tools release path works.  I broke the :latest release path but hopefully the current patch fixes that.

k2cli release path is looking good.  sorting out github secrets then just bugfixes.

**Comment by coffeepac** [comment]
fully into bugfix mode.  will have to wait until Monday to get everything merged but I'm feeling pretty good.

**Comment by coffeepac** [comment]
continuing to bugfix.  found some issues with the underlying image and hit the "null" check for k2cli that I hit in k2 & k2-tools.  

also hit a weird issue with github-release.  our makefile assumed we were on a branch, but jenkins checks out a specific rev in headless mode so had to add a parameter to specify the branch.  woo.

**Comment by coffeepac** [comment]
spent two more days debugging then realized I wasn't passing the branch name through the jenkinsfile into the makefile.  barf.  new PR up.

**Comment by coffeepac** [comment]
Jenkins jobs are complete and I was able to release a new version successfully.  Yay.

ToDo:
 - document how to cut a release
   - where to put document.  docs repo?
 - make sure to add manual brew update step

**Comment by coffeepac** [comment]
ToReview:
- follow the [instructions](https://github.com/samsung-cnct/docs/tree/master/cnct/common-tools/kraken-release.md) to release all three kraken components and update the homebrew tap

**Comment by leahnp** [comment]
Review status: 

- Kraken-tools: v0.4 tagged/pushed Jenkins failing due to ELB name length, need to rename Jenkins task
- Kraken-lib: v0.1 tagged/pushed Jenkins failing due to: `An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ImportError: No module named semver`

Some changes to the docs instructions (PR once this review is done)
- if pushing from fork you need to specify a remote. 
- need to specify commit hash to tag specific commit else latest is tagged
- some general tag visibility commands

About to start Kraken/Kraken homebrew

**Comment by leahnp** [comment]
Renaming jenkins jobs to:

- kraken-lib-rel
- kraken-tools-rel
- kraken-rel

Updating in docs as well.

**Comment by leahnp** [comment]
kraken-lib-rel & kraken-tools-rel releases cut successfully! 

I am a bit confused by the kraken release, I specified v1.1.0 with kraken-lib version v0.1 in the Jenkins console, it ran successfully: https://common-jenkins.kubeme.io/job/kraken-rel/23/ but no new releases were pushed to github - am I missing a step? @coffeepac

**Comment by coffeepac** [comment]
The release was posted to https://github.com/samsung-cnct/kraken-old/releases which is where https://github.com/samsung-cnct/kraken/releases resolves too.  

We will jeed to back this release out, rename the repos and then try it again.

**Comment by leahnp** [comment]
-removed incorrect release from kraken-old

@coffeepac not exactly sure which repos you want me to rename, will wait until you're into the office for the renaming

**Comment by leahnp** [comment]
-k2cli repo renamed to kraken, successfully cut 1.1.0
-https://github.com/samsung-cnct/homebrew-k2cli/pull/9
-https://github.com/samsung-cnct/docs/pull/34

**Comment by leahnp** [comment]
@alika to backup review:

Check new releases for our 3 kraken artifacts were released.
Check kraken homebrew was udpated with the new release
Check out samsung-cnct/docs#34 and make sure it makes sense
Close this issue and move to "DONE" column

**Comment by alika** [comment]
- check kraken release artifacts
  - checked https://github.com/samsung-cnct/kraken-tools/releases/tag/v0.4
  - checked https://github.com/samsung-cnct/kraken-lib/releases/tag/v0.1
  - checked https://github.com/samsung-cnct/kraken/releases/tag/1.1.0
- check kraken homebrew
  - ran `brew install kraken`
```
==> Installing samsung-cnct/k2cli/kraken
==> Downloading https://github.com/samsung-cnct/kraken/releases/download/1.1.0/kraken-rel_1.1.0_darwin_amd64.tar.gz
==> Downloading from https://github-production-release-asset-2e65be.s3.amazonaws.com/72028351/cbce5e6e-9a1a-11e7-9ad7-90da13db
######################################################################## 100.0%
🍺  /usr/local/Cellar/kraken/64: 3 files, 11.3MB, built in 42 seconds
$ kraken
Kraken is an orchestration and cluster level management system for Kubernetes that creates a production scale cluster on a range of platforms using its default settings.

Usage:
  kraken [command]

Available Commands:
  cluster     Kraken cluster actions
  docs        Generate markdown docs
  generate    Generate a Kraken config file
  release-it  Release the kraken!
  tool        Use tools with Kraken cluster
  version     Display cli version

Flags:
  -d, --docker-host string   Docker host address (default "unix:///var/run/docker.sock")
  -i, --image string         Krakenlib container image (default "quay.io/samsung_cnct/k2:v0.1")
  -a, --keep-alive           keep stopped containers.
  -k, --kraken string        Path to kraken config file (default "/Users/alika/.kraken/.kraken/config.yaml")
  -w, --log-path string      Save output output of container action to path
  -x, --log-success          Display full action logs on success
  -o, --output string        Kraken output folder (default "/Users/alika/.kraken")
  -t, --timeout int          timeout (in seconds) for container actions (default 1200)
      --tls                  Use TLS with the remote API
      --tlscacert string     Trust certs signed only by this CA (default "/Users/alika/.docker/ca.pem")
      --tlscert string       Path to the TLS certificate file (default "/Users/alika/.docker/cert.pem")
      --tlskey string        Path to the TLS key file (default "/Users/alika/.docker/key.pem")
      --tlsverify            Use TLS and verify the remote API
  -v, --verbosity            Verbose output

Use "kraken [command] --help" for more information about a command.
```
- check [kraken release docs](https://github.com/samsung-cnct/docs/pull/34/files)
  - verified links
  - for consistency the doc should use kraken-lib in the section markdown section header
